improvement: S3UTILS-79 document followerDiff scan procedure

jonathan-gramain opened the pull request:
In README.md, provide a procedure to run a full scan to detect
metadata discrepancies across active stateful members.

jonathan-gramain committed:
improvement: S3UTILS-79 document followerDiff scan procedure

In README.md, provide a procedure to run a full scan to detect
metadata discrepancies across active stateful members.

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/s3utils/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

watsaqat commented on README.md:
```diff
@@ -465,6 +465,217 @@ node followerDiff.js
 
+## Scan Procedure
+
+In order to scan a set of raft sessions to detect discrepancies
```
iirc, the use-case is to detect leader-follower discrepancies, correct? If so, lets mention that

jonathan-gramain replied:
I'd say that the *goal* is to detect any discrepancy between members, and the *means* is checking follower-leader discrepancies, due to how technically this is achievable without introducing downtime.

jonathan-gramain replied:
I'll add an extra paragraph to explain the general idea of the procedure.

watsaqat commented on README.md:
````diff
@@ -465,6 +465,217 @@ node followerDiff.js
+```
+
+Check that what you get looks like the above, then execute those
+commands on the shell (e.g. by piping with `| bash -x`).
````
maybe adding the full command to be run is better here

jonathan-gramain replied:
Agree, I can add the command to run

watsaqat commented on README.md:
```diff
@@ -465,6 +465,217 @@ node followerDiff.js
+- The current version of the script does not work with Metadata bucket
+  format v1, only with v0. If the need arises, it could be made to
+  work with v1 with some more work.
+
```
a few Qs:
* if any of the steps fail, should the user start over from step 1, 2, etc?
* will the user always know when a step has failed? could failures be silent at any of the steps or will timeouts alert the user?

jonathan-gramain replied:
There could be ways to re-do a single step that fails, either by just relaunching it or by doing a first cleanup before re-launching the step. I should give some details about those cases indeed.

I'd need to make sure the provided commands and scripts fail in a clear way (like checking the error messages or exit code).

This also makes me think that even before running the set of steps, there are pre-conditions to be checked, like making sure followers are up-to-date with the leader for example.

tmacro commented on README.md:
```diff
@@ -465,6 +465,217 @@ node followerDiff.js
+
+Finally, once all scans have been done on all active stateful hosts,
+gather all diff results for later analysis.
+
```
Could include a bit about what to expect in the results. Mine are empty files which I assume means that there are no differences. No need to get too technical just something high level about the different cases.

jonathan-gramain replied:
I can add a pointer to where the output format is described above in the `followerDiff` tool description. Indeed empty files mean no difference (one difference per line)

jonathan-gramain committed:
S3UTILS-79 address comments in followerDiff scan procedure

- Add a paragraph to describe the general idea of the scan

- Add a pre-check section to check that followers are up-to-date with
  the leader

- Give an example command to generate the list of raft sessions to
  scan on each server

- A few extra precisions regarding command usage, how to resume after
  error and simplifications

- Give a pointer to the diff output format

jonathan-gramain commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.13`



The following branches will **NOT** be impacted:

* `development/1.12`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.13`



The following branches have **NOT** changed:

* `development/1.12`



Please check the status of the associated issue S3UTILS-79.

Goodbye jonathan-gramain.
